x/auth/middleware/snapshot.go

```diff
@@ -0,0 +1,66 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+type snapshotHandler struct {
+	next tx.Handler
+}
+
+// SnapshotTxMiddleware branch out MultiStore in DeliverTx and commit if no error returned.
+func SnapshotTxMiddleware(txh tx.Handler) tx.Handler {
+	return snapshotHandler{next: txh}
+}
+
+// CheckTx implements tx.Handler.CheckTx method.
+func (sh snapshotHandler) CheckTx(ctx context.Context, req tx.Request, checkReq tx.RequestCheckTx) (tx.Response, tx.ResponseCheckTx, error) {
+	// Do nothing during CheckTx.
+	return sh.next.CheckTx(ctx, req, checkReq)
+}
+
+// DeliverTx implements tx.Handler.DeliverTx method.
+func (sh snapshotHandler) DeliverTx(ctx context.Context, req tx.Request) (tx.Response, error) {
+	// Create a new Context based off of the existing Context with a MultiStore branch
+	// in case message processing fails. At this point, the MultiStore
+	// is a branch of a branch.
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	runMsgCtx, msCache := cacheTxContext(sdkCtx, req.TxBytes)
+
+	rsp, err := sh.next.DeliverTx(sdk.WrapSDKContext(runMsgCtx), req)
+	if err != nil {
+		msCache.Write()
+	}
+
+	return rsp, err
+}
+
+// SimulateTx implements tx.Handler.SimulateTx method.
+func (sh snapshotHandler) SimulateTx(ctx context.Context, req tx.Request) (tx.Response, error) {
```

Comment:
we also need to cacheTxContext here i think. Or else simulation results can be wrong too

Reply:
thanks, done.